### DISCUSSION TITLE
Can I create my own UI?

### DISCUSSION THREAD

**Discussion by ivan-diachenko-td** [sparkle-project/Sparkle#2515] [Q&A]
I want to disable all Sparkle-related UI in my application and implement my own UI or install an update programmatically when the one is found. So do I understand it correctly that for example if I want to create my own UI to show that a new update is available, I need to implement `showUpdateFoundWithAppcastItem`? If I want to disable other modal dialogs I just need to override all `SPUStandardUserDriver` methods, right?

```obj-c

@interface SUAppUserDriver : SPUStandardUserDriver
@end

@implementation SUAppUserDriver

- (void)showUpdateFoundWithAppcastItem:(SUAppcastItem*)appcastItem
                                 state:(SPUUserUpdateState*)state
                                 reply:(void (^)(SPUUserUpdateChoice))reply {

  // Show my own UI
  reply(SPUUserUpdateChoiceInstall);
}

- (void)showReadyToInstallAndRelaunch:(void (^)(SPUUserUpdateChoice))reply {
  reply(SPUUserUpdateChoiceInstall);
}

- (void)showUpdaterError:(NSError *)error acknowledgement:(void (^)(void))acknowledgement {
  // Do nothing
}

// And so on
@end
```

Thanks in advance

**Comment by zorgiepoo** [comment]
`SPUUserDriver` is for creating your own UI, but it's not for disabling UI or for installing updates automatically. I don't recommend subclassing SPUStandardUserDriver. If you want to create your own UI, create your own.

See [this post](https://github.com/sparkle-project/Sparkle/discussions/2452#discussioncomment-7278412) for more context (though probably ignore the first paragraph about the pkg `installer`). If you actually want silent automatic updates, use SUAutomaticallyUpdate as referred to in that post. There is also a [delegate method](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:willInstallUpdateOnQuit:immediateInstallationBlock:) for handling updates being installed in the background. Note the user driver can still be shown in some cases, like when authorization is required or the user hasn't quit the app in a week (if the delegate method isn't overridden).

Try running this and see if it does what you expect:

```sh
sudo chown root /Applications/MyApp.app
```

..since you're probably not considering the case where user authorization is required to install an app. It is a common enough case to care about.

**Reply by ivan-diachenko-td** [reply]
@zorgiepoo Thank you so much for the response)). I really appreciate it. What I try to do is create my own API on top of Sparkle API (I am not interested in showing any Sparkle UI).  Then my clients would be able to use my API to create their own UI for example. Here is a pseudo-code (my API's entry point) I try to achieve:

```cpp
update_checker->checkForUpdates("http://localhost/app_updates/appcast.impl", [](const UpdatePackage& updatePackage, UpdateReply reply) {
    LOG(INFO) << "New Version Found: " << updatePackage.version;
     if (some_condition) {
         reply.install();
     } else {
         reply.skip();
     }   
    });
```

Is it possible to implement this approach with Sparkle API? Especially, I am interested in programmatic updates.

I am looking forward to hearing from you.

**Reply by zorgiepoo** [reply]
Create a custom SPUUserDriver and forward those events to your clients which will need to show their own UI. Clients failing to show UI and automating installation, especially in `showUpdateFoundWithAppcastItem:state:reply:` can lead to consequences like I outlined in my prior post which I am not sure you understood.

Be sure you understand Sparkle's default behavior in how it checks for updates from the website documentation [testing section](https://sparkle-project.org/documentation#6-test-sparkle-out) and [customizing Sparkle's behavior](https://sparkle-project.org/documentation/customization/). By default developers do not issue checkForUpdates checks, Sparkle just does that automatically in the background on a periodic interval. For your use case it's possible you may want to disable that.

Also Sparkle has another "automatic download/install of updates silently in the background" mode which defers using the user driver (silent installs are not always possible like in the authorization case), but presumably you may not want that. However if you want the update to be downloaded automatically before it's shown to the user via the user driver, then this may be needed as well.
